feat(core): support adding additional cache key to cdk context values

luxaritas opened the pull request:
The `ContextProvider` mechanism and various "lookup" functions of a number of constructs support caching resolved values in the cdk.context.json. The context keys are constructed from the parameters of the lookup, which for lookup functions means whenever a resource with the same parameters is resolved, it is resolved as the same value across the entire app. However when a value may change over time, the user may wish to use the latest value when creating creating a new reference to the construct, effectively tying the cached context value to the scope - this patch enables this.

The primary use case is looking up an AMI parameter for a "stateful" EC2 instance. Currently if you specify `cachedInContext`, any future images created would use the same cached AMI, and updating the value would require updating all usages of the image across the entire app.

Closes #26982

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation reviewed (dismissed):
The pull request linter has failed. See the aws-cdk-automation comment below for failure reasons. If you believe this pull request should receive an exemption, please comment and provide a justification.


A comment requesting an exemption should contain the text `Exemption Request`. Additionally, if clarification is needed add `Clarification Request` to a comment.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 8268fc8774923daeec36721ce96d041818e098f9
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7d807172-9798-45a9-93c6-5949e324e551%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

luxaritas commented:
Some open questions/areas I am particularly looking for feedback (in particular being new to contributing here):
* I added a new optional parameter wherever I extended a lookup function. This was the simplest and most consistent backward-compatible option, however I don't find it a particularly elegant API (whereas most other areas of the API are really good about throwing miscellaneous options in options objects for extensibility). What's the preferred approach here?
* I added the option to all lookups that called `ContextProvider.getValue`, aside from AZ lookups where I felt it particularly didn't make sense. Was this the correct call? Would it be better to limit this just to EC2 instances and parameters which are more likely to have a concrete use case for this (where different values across time are useful concurrently)?
* I added tests where there were already context-related tests that I could find, as well as an integration test for a machine image lookup which is the most complex out of the largely-identical situations created here. Should I add tests elsewhere?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 318dfe83a7033d9d04629263a2a55d4494521411
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=cf584e43-14bc-4b1a-a3fc-ad5c7419f37e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

luxaritas commented:
One more potential question: should `linkContextToScope` optionally take a `Construct` instead of a boolean so that it could be linked to a specific scope as opposed to the scope the value is directly used in? Though given the inability of unions to be used per the design guidelines, I'm not sure what the ideal way to model that would be...

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 802f8ba047713c04e388083bb17c467a994514b0
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=89977c86-dd2b-4fd6-963d-969c79ad1d52%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: d6f96bd61bc607da88a6cea94d2fd8793cb8ffac
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=53d1a721-98d4-4b76-85e0-af620e0a5005%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: edbccad93b47218fd0ae0de269b487d371c2bc7a
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bc145372-dc1a-4b13-b2e2-acf1e0639b4a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

rix0rrr reviewed (changes requested):
Hey, thanks for the idea and the contribution!

I have to say that I'm reflexively experiencing an aversion to this proposed change 🤢😝 because why would those values ever be different? But I can also appreciate that this is necessary to evolve a stateful app over time.

The solution of including the scope is clever... but I wonder if it doesn't give more flexibility at the cost of a vague contract.

Could we do something blindingly obvious instead, have the user include a cache busting key? And then everyone can reuse/invalidate that key on a per-location basis as they desire? 

```ts
ec2.MachineImage.latestAmazonLinux2023({
  cachedInContext: true,

  additionalCacheKey: 'asdf',  // Can change this to qwerty whenever I feel like updating it
});
```

They can still make that be the scope if they want to (achieving the current behavior) by passing `this.node.path`.

luxaritas commented:
Hey @rix0rrr! I do think that makes sense, and now that you say that I realize that did feel a bit weird. I imagine that cache key would be used as the value of a prop passed to the context, I'm thinking maybe call they key of that prop `userCacheKey`. I'll adjust the PR shortly and we can continue discussion from there

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 285fa901d0a0d8a27ceeb4169498ae4666e2f507
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c885de3b-9103-4226-9649-ce7c9056f161%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: bf8c6c1ad97db69866321fcffeaf9adb4f943972
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d4c5128d-98c0-4511-8e9b-f061adb6c6ae%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

luxaritas commented:
I also wound up moving additionalCacheKey into the lookup props - it seems like it wound up being cleaner. I'll adjust the PR title and description to match the change

lpizzinidev reviewed (changes requested):
Thanks (and sorry for the late review) 👍
I left you some comments as a first revision.

luxaritas commented on packages/aws-cdk-lib/aws-ec2/lib/machine-image/common.ts:
```diff
@@ -66,12 +72,14 @@ export interface AmazonLinuxImageSsmParameterBaseProps extends AmazonLinuxImageS
 
 export abstract class AmazonLinuxImageSsmParameterBase implements IMachineImage {
   private readonly cachedInContext: boolean;
+  private readonly additionalCacheKey?: string;
   constructor(private readonly props: AmazonLinuxImageSsmParameterBaseProps) {
     this.cachedInContext = this.props.cachedInContext ?? true;
+    if (props.additionalCacheKey) this.additionalCacheKey = this.props.additionalCacheKey;
```
This seems to be true of cachedInContext as well (which I emulated) - should I change that one as well?

luxaritas replied:
Never mind, forgot that one has a default set

luxaritas commented on packages/aws-cdk-lib/aws-ec2/lib/machine-image/machine-image.ts:
```diff
@@ -652,11 +667,16 @@ export class GenericWindowsImage implements IMachineImage {
  * will be used on future runs. To refresh the AMI lookup, you will have to
  * evict the value from the cache using the `cdk context` command. See
  * https://docs.aws.amazon.com/cdk/latest/guide/context.html for more information.
+ * If `props.additionalCacheKey` is true, the context key use that value as a discriminator
+ * rather than the cached value being global across all lookups
```
Other similar docstrings have "so that separate lookups with the same parameters can have separate cache lifecycles" - should we just remove that extra language here, use the same language as elsewhere here, or do you think it needs to be removed elsewhere?

luxaritas commented on packages/aws-cdk-lib/core/lib/context-provider.ts:
```diff
@@ -67,9 +67,10 @@ export class ContextProvider {
-    const props = options.includeEnvironment ?? true
-      ? { account: stack.account, region: stack.region, ...options.props }
-      : (options.props ?? {});
+    const props = {
+      ...((options.includeEnvironment ?? true) ? { account: stack.account, region: stack.region } : {}),
+      ...options.props,
+    };
```
I agree, may have been an unintentional intermediate change

luxaritas commented on packages/aws-cdk-lib/aws-ec2/lib/vpc.ts:
```diff
@@ -1292,6 +1292,7 @@ export class Vpc extends VpcBase {
         returnAsymmetricSubnets: true,
         returnVpnGateways: options.returnVpnGateways,
         subnetGroupNameTag: options.subnetGroupNameTag,
+        ...(options.additionalCacheKey ? { additionalCacheKey: options.additionalCacheKey } : {}),
```
Just noticed that undefined values are stripped in context-provider. `props` is just `{ [key: string]: any }`, so I think there's nothing to do beyond that?

luxaritas replied:
...never mind, I see the interface this is being specified as.

luxaritas commented on packages/aws-cdk-lib/aws-ecs/lib/amis.ts:
```diff
@@ -385,8 +410,8 @@ Object.defineProperty(BottleRocketImage.prototype, BR_IMAGE_SYMBOL, {
 });
 
-function lookupImage(scope: Construct, cachedInContext: boolean | undefined, parameterName: string) {
+function lookupImage(scope: Construct, cachedInContext: boolean, parameterName: string, additionalCacheKey?: string) {
```
I think this was a result of an intermediate change.

Out of curiosity: This appears to be an "internal" function that is never called with undefined - does this still have a backward-compat issue just by virtue of it being exported?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: f7bcd97c2674dcda508ff9b97be2a78ec51f0f15
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=98b532c0-d984-4a72-a9e2-f7260d6b907d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 8c508ab0f7c58996544cc7a9056cff480dd7a30e
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=cd948b04-9f52-4f96-821c-6225bba7238d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

lpizzinidev commented on packages/aws-cdk-lib/aws-route53/README.md:
```diff
@@ -247,6 +247,14 @@ route53.HostedZone.fromLookup(this, 'MyZone', {
+called `cdk.context.json`. You must commit this file to source control so
+that the lookup values are available in non-privileged environments such
+as CI build steps, and to ensure your template builds are repeatable.
+To customize the cache key use the `additionalCacheKey` parameter.
+This can be useful if you want to scope the context variable to a construct 
+(eg, using `additionalCacheKey: this.node.path`).
+
```
```suggestion

To customize the cache key use the `additionalCacheKey` parameter.
This can be useful if you want to scope the context variable to a construct 
(eg, using `additionalCacheKey: this.node.path`).

```ts
route53.HostedZone.fromLookup(this, 'MyZone', {
  domainName: 'example.com',
  additionalCacheKey: 'additionalKey',
});

lpizzinidev commented on packages/aws-cdk-lib/aws-ssm/README.md:
```diff
@@ -58,6 +58,15 @@ This method uses AWS API calls to lookup the value from SSM during synthesis.
+called `cdk.context.json`. You must commit this file to source control so
+that the lookup values are available in non-privileged environments such
+as CI build steps, and to ensure your template builds are repeatable.
+To customize the cache key use the `additionalCacheKey` parameter.
+This can be useful if you want to scope the context variable to a construct 
+(eg, using `additionalCacheKey: this.node.path`).
+
+
```
```suggestion

To customize the cache key use the `additionalCacheKey` parameter.
This can be useful if you want to scope the context variable to a construct 
(eg, using `additionalCacheKey: this.node.path`).

```ts
const stringValue = ssm.StringParameter.valueFromLookup(this, '/My/Public/Parameter', this.node.path);
```

lpizzinidev reviewed (changes requested):
Thanks for the changes.
I left some notes on possible documentation improvements.

> Also given the cloud-assembly-schema is now conflicting, let me know if I should merge or rebase to latest main and regen that

You should merge `main` into the PR branch.

aws-cdk-automation commented:
This PR cannot be merged because it has conflicts. Please resolve them. The PR will be considered stale and closed if it remains in an unmergeable state.

luxaritas commented on packages/aws-cdk-lib/aws-ec2/lib/machine-image/machine-image.ts:
```diff
@@ -728,5 +746,11 @@ export interface LookupMachineImageProps {
   readonly userData?: UserData;
+
+  /**
+   * When the image is cached in cdk.context.json, adds an additional discriminator to the
+   * cache key so that separate lookups with the same parameters can have separate cache lifecycles
```
For clarity I'm thinking "If cachedInContext is true, adds an additional discriminator..." (since this variable does nothing on its own)

luxaritas replied:
Hm, unfortunately cachedInContext does not always appear on the lookup parameters object. Probably better not to do it for the sake of consistency then. If anything it would be "adds an additional discriminator... if cached", but not sure how helpful that really is at that point

luxaritas commented on packages/aws-cdk-lib/aws-elasticloadbalancingv2/README.md:
```diff
@@ -519,6 +519,14 @@ following lookup methods:
 - `NetworkListener.fromLookup(options)` - Look up a network load balancer
   listener.
 
+The result of a `fromLookup()` operation will be written to a file
```
The general description is applicable to both load balancers and load balancer listeners so I think that should be in the top-level section (and I'll fix the heading level for LBL lookup options while I'm at it). Will add the examples

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 8397e442ebfaaeb99e8a5affb9ad2d3be69486fa
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e52b7a33-37e1-4cd9-9dce-ae8bae32be2f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 4cdc802c9f4bbfecb98d0ed634321e7cdcf313e4
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a504307c-3833-4708-adec-127c9b866753%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

luxaritas commented:
For some reason the cloud-assembly-schema as I generated locally appears to have some items reordered vs when run via CI...? Not sure how to resolve, I ran `yarn update-schema` and its still failing

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 1b6955c99c3fca254bd7f82f629c2625869416fa
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4f72377d-e62a-4504-becc-4d8aa7ed6929%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

rix0rrr commented on packages/aws-cdk-lib/aws-ecs/README.md:
````diff
@@ -156,6 +156,19 @@ const autoScalingGroup = new autoscaling.AutoScalingGroup(this, 'ASG', {
 });
 ```
 
+To customize the cache key use the `additionalCacheKey` parameter.
````
You need to explain WHY a user might want to do this. If I didn't know what this was already, reading this bit of the documentation certainly wouldn't help me.

rix0rrr commented on packages/aws-cdk-lib/aws-elasticloadbalancingv2/README.md:
````diff
@@ -548,7 +557,18 @@ const loadBalancer = elbv2.ApplicationLoadBalancer.fromLookup(this, 'ALB', {
+```ts
+const loadBalancer = elbv2.ApplicationLoadBalancer.fromLookup(this, 'ALB', {
+  loadBalancerArn: 'arn:aws:elasticloadbalancing:us-east-2:123456789012:loadbalancer/app/my-load-balancer/1234567890123456',
+  additionalCacheKey: this.node.path,
````
```
// This will ensure that ...(what)...?
```

rix0rrr reviewed (changes requested):
Almost there! You did a good job updating the docs explaining WHAT people can now do, but what's missing is WHY they would want to do this.

Add that and it's golden.

rix0rrr commented on packages/aws-cdk-lib/cloud-assembly-schema/lib/cloud-assembly/context-queries.ts:
```diff
@@ -92,6 +92,13 @@ export interface AmiContextQuery {
+   *
+   * @default - no additional cache key
+   */
+  readonly additionalCacheKey?: string;
```
When the context provider sends out the query, it also picks the context cache key. We can just mix the additional string into the context key; it never needs to make it to the cloud assembly as a separate field.

luxaritas replied:
I believe the reason I did this was so that when the queries are constructed like so:

```ts
const attributes: cxapi.KeyContextResponse = ContextProvider.getValue(scope, {
      provider: cxschema.ContextProvider.KEY_PROVIDER,
      props: {
        aliasName: options.aliasName,
        additionalCacheKey: options.additionalCacheKey,
      } as cxschema.KeyContextQuery,
      ...
});
```
it would be surfaced as a valid part of props. You're saying this is not necessary?

aws-cdk-automation commented:
<!--STALE PR-->
This PR has been in the CHANGES REQUESTED state for 3 weeks, and looks abandoned. To keep this PR from being closed, please continue work on it. If not, it will automatically be closed in a week.

luxaritas commented:
@rix0rrr Could you provide clarification on the comment above so that I can finish this up? Thanks!

aws-cdk-automation commented:
This PR has been deemed to be abandoned, and will be automatically closed. Please create a new PR for these changes if you think this decision has been made in error.
